src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/X86Base.cs

```diff
@@ -44,6 +45,18 @@ public abstract class X64
             /// Its functionality is exposed in the public <see cref="System.Numerics.BitOperations" /> class.
             /// </remarks>
             internal static ulong BitScanReverse(ulong value) => BitScanReverse(value);
+
+            /// <summary>
+            /// unsigned __int64 _udiv128(unsigned __int64 highdividend, unsigned __int64 lowdividend, unsigned __int64 divisor, unsigned __int64* remainder)
+            ///   DIV reg/m64
+            /// </summary>
+            public static (ulong Quotient, ulong Remainder) DivRem(ulong lower, ulong upper, ulong divisor) => DivRem(lower, upper, divisor);
```

Comment:
I'd expect a few managed methods, such as `Int32/Int64/Math/UInt32/UInt64.DivRem` to be updated to use this handling or be likewise directly imported as this intrinsic (as part of this PR).

Long term I'd expect we also add an opt to `morph` to recognize `/` + `%` and optimize down to `DivRem`, but that's likely out of scope of this PR.

Reply:
I had a discussion with @pentp in https://github.com/dotnet/runtime/pull/66551#issuecomment-1398928443 and offline and we decided to not do it in this PR.

Reply:
I think it's better to go ahead and utilize this anyways, particularly for the `Int32/Int64/UInt32/UInt64` cases that have identical signatures. That will help improve test coverage, increase confidence in correctness, and can be used to help drive fixing the other issue if it's a big enough of an impact.

Reply:
> That will help improve test coverage, increase confidence in correctness, and can be used to help drive fixing the other issue 

That is definitely an advantage. @pentp ?

Reply:
I'm also in favour of using it for `Math.DivRem` and similar, but that would require solving some additional issues that we felt would be best addressed in a separate follow-up PR. Namely the constant divisor case would need to utilize `RuntimeHelpers.IsKnownConstant` to bypass the intrinsic (and new overloads need to be added for `IsKnownConstant`) and the regular divisor cases had some occasional register allocation issues also that caused unnecessary memory copies. Plus there's the issue of forfeiting use of `CDQ/CQO` that's probably not easily solved at C# level and would instead need some special handling at the importer (or something like that).

So I would like to see the intrinsic itself merged and then get the follow-up use cases as smaller PRs.

Reply:
What is tracking this work being done as part of .NET 8

I don't think we can/should check in the API if it's being exposed in such a way that it is worse than public API for the typical scenario. That will lead developers down the wrong path for an API that is meant to be oriented towards performance scenarios.

Based on the above, most of these look like trivial 1-2 cycle differences that may end up balancing out for other reasons. I don't see anything that indicates the cost increase is significant or should block its broader use.

Making it more used now will also cause any necessary additional work, such as what appear to be constant folding support and basic containment of casts, to be prioritized as part of .NET 8.
-------------

> Namely the constant divisor case would need to utilize RuntimeHelpers.IsKnownConstant to bypass the intrinsic (and new overloads need to be added for IsKnownConstant)

I don't think this is needed. We just need to ensure the relevant constant folding support exists for `NI_X86Base_DivRem` instead. `IsKnownConstant` has known issues/limitations and increases the size of the imported IR which can have other negative consequences/limits.

> the regular divisor cases had some occasional register allocation issues that caused unnecessary memory copies.

This doesn't seem like a blocker and is something that any code can hit. It sounds like something we can either live with or push towards trivially fixing separately. At a glance, I believe that this can be trivially handled with a couple tweaks in codegen and the lsra building for this intrinsic.

> Plus there's the issue of forfeiting use of CDQ/CQO that's probably not easily solved at C# level and would instead need some special handling at the importer (or something like that).

This should really just be a special case of processing (containing) `CAST` as part of the HWIntrinsic, it should likewise not be a blocker to the feature being merged and used more generally.

Reply:
> I'd expect a few managed methods, such as `Int32/Int64/Math/UInt32/UInt64.DivRem` to be updated to use this handling or be likewise directly imported as this intrinsic (as part of this PR).

I had a prototype branch on my another machine, for `Decimal`. I planned to put off it after this PR.